### PR TITLE
Ingest shaper

### LINKED ISSUE #1833: Add facility to run Z scripts at ingest
Add a facility to run Z scripts on ingested data, before it lands. This should also be applicable to the pcap ingest path.

### PR DESCRIPTION
**Pull request by henridf**
This commit adds support for a running Z "shaper" script on incoming logs. This facility is intended for ETL tasks such as transforming record layouts and field names, casting leaf types, or policing incoming types against a set of admissible schemas. (Z script does not quite yet have the support for much of this, but that is imminent).

Closes #1833

### CODE REVIEW

File: api/client/multipartwriter.go [hunk 1]
```diff
@@ -125,6 +135,17 @@ func (m *MultipartWriter) sendJSONConfig() error {
...
+	if m.shaper == nil {
+		return nil
+	}
+	w, err := m.form.CreateFormField("shaper_ast")
+	if err != nil {
+		return err
+	}
+	return json.NewEncoder(w).Encode(m.shaper)
+}
+
```
Comment (mattnibs) [inline api/client/multipartwriter.go:148]: Can we drop the `sendJSONConfig` or would this be done later?

File: api/client/connection.go [hunk 2]
```diff
@@ -368,11 +370,18 @@ func (c *Connection) LogPostPath(ctx context.Context, space api.SpaceID, opts *L
 	return payloads.Error()
 }
 
-func (c *Connection) LogPostPathStream(ctx context.Context, space api.SpaceID, opts *LogPostOpts, paths ...string) (*Stream, error) {
-	body := api.LogPostRequest{Paths: paths}
-	if opts != nil {
-		body.JSONTypeConfig = opts.JSON
-		body.StopErr = opts.StopError
+func (c *Connection) LogPostPathStream(ctx context.Context, space api.SpaceID, opts LogPostOpts, paths ...string) (*Stream, error) {
```
Comment (mattnibs) [inline api/client/connection.go:373]: I would just assume keep this option a pointer so you don't have to do `api.LogPostOpts` in the vanilla post case.

File: api/client/connection.go [hunk 3]
```diff
@@ -368,11 +370,18 @@ func (c *Connection) LogPostPath(ctx context.Context, space api.SpaceID, opts *L
...
+	}
+	if opts.Shaper != nil {
+		raw, err := json.Marshal(opts.Shaper)
+		if err != nil {
+			return nil, err
+		}
+		body.Shaper = raw
 	}
 	req := c.Request(ctx).
 		SetBody(body)
```
Comment (mattnibs) [inline api/client/connection.go:387]: Shouldn't the methods `LogPost`, `LogPostReaders` and `LogPostWriter` also handle the `Shaper` option?

File: cmd/zapi/cmd/post/post.go [hunk 4]
```diff
@@ -55,6 +55,9 @@ func (c *PostCommand) Run(args []string) (err error) {
 	if err != nil {
 		return err
 	}
+	if c.postFlags.shaperAST != nil {
+		c.logwriter.SetShaper(c.postFlags.shaperAST)
+	}
```
Comment (mattnibs) [inline cmd/zapi/cmd/post/post.go:60]: Perhaps in this case create a `LogPostOpts` with the ast and pass this into `LogPostWriter`?

File: ppl/zqd/ingest/multipartlogreader.go [hunk 5]
```diff
@@ -46,6 +53,12 @@ read:
 		if zr == nil || err != nil {
 			return nil, err
 		}
+		if m.shaperAST != nil {
+			zr, err = driver.NewReader(context.Background(), m.shaperAST, m.zctx, zr)
+			if err != nil {
+				return nil, err
+			}
+		}
```
Comment (mattnibs) [inline ppl/zqd/ingest/multipartlogreader.go:61]: Consider moving this into the `next` function that creates the reader.

**Review by mattnibs — COMMENTED** [review]
This great! Can't wait to use this on the timestamp transform for suricata!

**Review by mattnibs — APPROVED** [review]

File: api/client/multipartwriter.go (hunk 1 above)
Comment (henridf) [reply api/client/multipartwriter.go:148]: Hmm, I'm assuming no one is using `sendJSONConfig` on multi-upload endpoint? Happy to remove right away if that's the case.

File: ppl/zqd/ingest/multipartlogreader.go [hunk 6]
```diff
@@ -57,14 +70,14 @@ read:
...
-			m.appendWarning(zr, err)
+			m.appendWarning(zr.(*zbuf.File), err)
 		}
 		goto read
 	}
 	return rec, err
 }
 
-func (m *MultipartLogReader) next() (*zbuf.File, error) {
+func (m *MultipartLogReader) next() (zbuf.ReadCloser, error) {
```
Comment (alfred-landrum) [inline ppl/zqd/ingest/multipartlogreader.go:80]: Can you explain why you needed to change this return type?

File: driver/driver.go [hunk 7]
```diff
@@ -214,9 +216,14 @@ func NewReaderWithConfig(ctx context.Context, conf Config, program ast.Proc, zct
...
 	}
-	return &muxReader{
+	mr := &muxReader{
 		cancel:      cancel,
 		muxOutput:   mux,
 		statsTickCh: make(chan time.Time),
-	}, nil
+		zr:          ioutil.NopCloser(nil),
+	}
+	if _, ok := reader.(io.Closer); ok {
```
Comment (alfred-landrum) [inline driver/driver.go:225]: I'm not seeing why the changes required added this type assertion, can you briefly explain?

File: ppl/zqd/ingest/multipartlogreader.go (hunk 6 above)
Comment (henridf) [reply ppl/zqd/ingest/multipartlogreader.go:80]: Because this might be passed to driver.NewReader which expects a zbuf.ReadCloser.

File: driver/driver.go (hunk 7 above)
Comment (henridf) [reply driver/driver.go:225]: The reader might be a `multipart.Part` which needs to be closed.
One alternate implementation would be to change the signature to pass a `zbuf.ReadCloser` instead, and do away with the type assertion.

File: cmd/zapi/cmd/post/post.go (hunk 4 above)
Comment (henridf) [reply cmd/zapi/cmd/post/post.go:60]: Absolutely! Not sure why I didn't do that in the first place.

File: api/client/multipartwriter.go (hunk 1 above)
Comment (mattnibs) [reply api/client/multipartwriter.go:148]: I would assume its being used (cc: @philrz). Isn't there some default jsonConfig we use for zeek logs and suricata logs? Perhaps wait until we have a shaper that replaces the legacy config?

File: api/client/connection.go (hunk 3 above)
Comment (henridf) [reply api/client/connection.go:387]: done in 085394f24905391eff81c5d9a26e0017284d9352

File: cmd/zapi/cmd/post/post.go (hunk 4 above)
Comment (henridf) [reply cmd/zapi/cmd/post/post.go]: done in 085394f24905391eff81c5d9a26e0017284d9352

File: api/client/connection.go (hunk 2 above)
Comment (henridf) [reply api/client/connection.go]: Done in 9cfce89f151ce9b9497e4ffa92f5215b6c1c2f4e.

(I think passing a zero-valued struct is preferable to passing a nil, but that's best kept to a separate change)

File: api/client/multipartwriter.go (hunk 1 above)
Comment (philrz) [reply api/client/multipartwriter.go:148]: It's not clear from the code fragment alone what user-facing functionality we might be talking about dropping here. If we're talking about something the Brim app uses, yes, there's built-in JSON typing config for out-of-the-box Zeek v3.x that is applied whenever a user attempts to import NDJSON (in fact, I have PR https://github.com/brimsec/brim/pull/1307 up right now for review that proposes an adjustment for it so we're better prepared for Zeek v4.x). We put that in place way back when we weren't sure how many users would bring-their-own-Zeek-TSV-logs vs. bring-their-own-Zeek-NDJSON-logs vs. bring-their-own-pcaps. By now we know that pcaps are by far what's been most popular, and at this point I can't say I'm aware of any community user that actually imported their own Zeek NDJSON logs. That doesn't mean it hasn't happened... one of the negatives of something being built-in and "just working" is that users might try it and succeed and not think twice about it. That said, knowing how brittle it is. I'd be willing to guess it's not regularly used, if ever. The one community user I _do_ know that went down the route of creating custom NDJSON type configs ended up doing it with `zq` outside of Brim to turn their NDJSON into ZNG and then importing _that_ into Brim, for example.

If it's not hurting anything to leave it alone until the ingest shaper has reached feature parity with what was there before, that might be the safest option.

Comment (henridf) [reply api/client/multipartwriter.go:148]: 👍 agreed

File: ppl/zqd/ingest/multipartlogreader.go (hunk 5 above)
Comment (henridf) [reply ppl/zqd/ingest/multipartlogreader.go]: done in 6a76767624ade2687273a0b2d3dbeef14d2a7673
